src/Ledger/TokenAlgebra.lagda

```diff
@@ -1,29 +1,43 @@
 \section{Token Algebra}
 \begin{code}[hide]
 {-# OPTIONS --safe #-}
-module Ledger.TokenAlgebra where
 
-open import Ledger.Prelude
 
+open import Agda.Primitive renaming (Set to Type)
+
+module Ledger.TokenAlgebra
+
+       -- TODO: decide/demonstrate how these parameters are to be defined in modules that depend on this one.
+       {PolicyID : Type}    -- identifies monetary policies
+       {ByteString : Type}  -- could postulate `ByteString` here, but then we'd have to drop `--safe` pragma
+       {AdaName : ByteString}
+ where
+
+open import Ledger.Prelude hiding (Coin ; Rel)
 open import Algebra using (CommutativeMonoid)
-open import Algebra.Morphism
-open import Data.Integer
+open import Algebra.Morphism using (IsCommutativeMonoidMorphism-syntax)
+open import Data.Integer using (ℤ)
 open import Data.Nat.Properties using (+-0-commutativeMonoid)
+open import Relation.Binary using (Rel)
+open import Relation.Unary using (Pred)
+
+record TokenAlgebra : Type 1ℓ where
+  field Value-CommutativeMonoid : CommutativeMonoid 0ℓ 0ℓ
 
-record TokenAlgebra : Set₁ where
-  field PolicyId : Set
-        Value-CommutativeMonoid : CommutativeMonoid 0ℓ 0ℓ
+  Coin : Type
+  Coin = ℕ
 
+  MemoryEstimate : Type
   MemoryEstimate = ℕ
 
   open CommutativeMonoid Value-CommutativeMonoid using (_≈_; ε)
        renaming (Carrier to Value; refl to reflᵛ; _∙_ to _+ᵛ_) public
 
   field coin     : Value → Coin
         inject   : Coin → Value
-        policies : Value → ℙ PolicyId
+        policies : Value → Pred PolicyID 0ℓ
```

Comment:
This won't work for the UTXOW rule, since we need decidable membership of `policies`.

Reply:
Yes, well, `Pred` is indeed moving toward a more general, Agda standard library-ish direction.  It's true it doesn't presuppose decidability. But should we not view that a feature, not a bug, since then decidability must be explicitly postulated (e.g., with a module parameter) if and when (and only if and when) it is needed?

Reply:
To be more specific, here's where `policies` will be used:
```
scriptsNeeded : UTxO → TxBody → ℙ ScriptHash
scriptsNeeded utxo txb =
  mapPartial ((λ { (inj₂ sh) → just sh ; _ → nothing }) ∘ payCred ∘ proj₁) ((utxo ˢ) ⟪$⟫ txins txb)
  ∪ mapPartial (λ { record { credential = inj₂ sh } → just sh ; _ → nothing }) (setFromList $ txvote txb)
  ∪ map policyIdToScriptHash $ policies (mint txb)
```
The only reason it's not there yet is that we don't have a function `policyIdToScriptHash` yet.